// components/Otp/Otp.js
import React from 'react';
import PropTypes from 'prop-types';

const pad = (n) => (n < 10 ? '0' + n : '' + n);

const computeDuration = (props) => {
    const minutes = typeof props.minutes === 'number' ? props.minutes : 0;
    const seconds = typeof props.seconds === 'number' ? props.seconds : 0;
    const total = Math.max(0, Math.floor(minutes * 60 + seconds));
    return total;
};

class Otp extends React.Component {
    constructor(props) {
        super(props);
        this.resendOTP = this.resendOTP.bind(this);
        this.tick = this.tick.bind(this);

        const duration = computeDuration(props);
        this.state = {
            start: Date.now(),
            duration: duration,
            minutes: pad(Math.floor(duration / 60)),
            seconds: pad(duration % 60),
            resend: duration === 0,
        };

        this.timerId = null;
    }

    componentDidMount() {
        if (this.state.duration > 0) {
            this.startTimer();
        }
    }

    componentDidUpdate(prevProps) {
        if (prevProps.minutes !== this.props.minutes || prevProps.seconds !== this.props.seconds) {
            this.reset();
        }
    }

    componentWillUnmount() {
        this.clearTimer();
    }

    clearTimer() {
        if (this.timerId) {
            clearInterval(this.timerId);
            this.timerId = null;
        }
    }

    startTimer() {
        this.clearTimer();
        this.timerId = setInterval(this.tick, 1000);
    }

    tick() {
        const elapsed = Math.floor((Date.now() - this.state.start) / 1000);
        const remaining = Math.max(0, this.state.duration - elapsed);
        const minutes = Math.floor(remaining / 60);
        const seconds = remaining % 60;

        this.setState({ minutes: pad(minutes), seconds: pad(seconds) });

        if (remaining <= 0) {
            this.clearTimer();
            this.setState({ resend: true });
            if (typeof this.props.onTimerComplete === 'function') {
                this.props.onTimerComplete();
            }
        }
    }

    reset() {
        const duration = computeDuration(this.props);
        this.setState(
            {
                start: Date.now(),
                duration: duration,
                minutes: pad(Math.floor(duration / 60)),
                seconds: pad(duration % 60),
                resend: duration === 0,
            },
            () => {
                if (duration > 0) this.startTimer();
                else this.clearTimer();
            }
        );
    }

    resendOTP(evt) {
        if (evt && evt.preventDefault) evt.preventDefault();
        if (!this.state.resend) return;
        this.reset();
        if (typeof this.props.resendEvent === 'function') {
            this.props.resendEvent();
        }
    }

    render() {
        const style = this.props.style || {};
        const seperator = this.props.seperator != null ? this.props.seperator : ' : ';
        const resendButtonText = this.props.resendButtonText || 'Resend';
        const timerLabel = this.state.minutes + seperator + this.state.seconds;

        return (
            <div>
                <h3 className="errMsg" style={style.otpTimer}>{timerLabel}</h3>
                <button
                    type="button"
                    className="resend"
                    style={style.resendBtn}
                    onClick={this.resendOTP}
                    disabled={!this.state.resend}
                >
                    {resendButtonText}
                </button>
            </div>
        );
    }
}

Otp.propTypes = {
    minutes: PropTypes.number,
    seconds: PropTypes.number,
    seperator: PropTypes.string,
    resendButtonText: PropTypes.string,
    resendEvent: PropTypes.func,
    onTimerComplete: PropTypes.func,
    style: PropTypes.shape({
        otpTimer: PropTypes.object,
        resendBtn: PropTypes.object,
    }),
};

Otp.defaultProps = {
    minutes: 0,
    seconds: 0,
    seperator: ' : ',
    resendButtonText: 'Resend',
    style: {},
};

export default Otp;
